### DISCUSSION TITLE
Correct setup of dependency injection for Windows Service & asp.net framework?

### DISCUSSION THREAD

**Discussion by ajohnstone-ks** [MarimerLLC/csla#1983] [Questions] [answered]
I'm a bit unclear on the proper configuration of Csla & the service provider when using the server data portal methods.

Currently we are doing this to set it up, both in a Windows Server application (which uses the Local data portal only) and on our Asp.net framework application which hosts the remote data portal our desktop clients use.

```
var container = new UnityContainer();
// Configure the container, including IDbConnection as a trainsient.
var provider = new Unity.Microsoft.DependencyInjection.ServiceProviderFactory(container);
ApplicationContext.DefaultServiceProvider = provider.CreateServiceProvider(container);
```

I think we are seeing some connection leaks though, as I've had to restart the Windows Service as it was getting timeouts getting a SqlConnection from the pool.  I think our remote data portal application server has the same issue as well, as things slow down over time and we need to restart it.

I think the issues crop up slowly over time as we've only just started converting things from the old way we got a connection to using `[Inject]` for our dependencies, but I'd like to verify that we are setting things up correctly so that, in the end, the `UnityContainer` which is created by the ServiceProvider is being deposed after every server side data portal operation.

Is the above correct, or am I missing something?

**Comment by rockfordlhotka** [comment]
CSLA doesn't manage the DI container _in any way_. It just uses the DI container.

In other words, whether the DI container is disposed or not is 100% up to the container (or you?).

**Reply by ajohnstone-ks** [reply]
Does Csla ever call `IServiceScopeFactory.CreateScope`?  If not, how is it getting the scope?

EDIT;

It looks like it does when you get `ApplicationContext.ServiceProviderScope` if the scope is null; I'm trying to trace where in Csla that is used.

Is there some place we're supposed to add code to create the scope instead of letting Csla do it for use in a remote data portal host application (asp.net), and then dispose it?

**Reply by rockfordlhotka** [reply]
Have you reviewed this issue? https://github.com/MarimerLLC/csla/issues/1874

The behavior changed in 5.4.0 to try and avoid creating a custom scope, because aspnet normally creates a per-request scope. So really you should not need to create your own scope, because aspnet should do that.

(that said, I'm finding that server-side Blazor is a different situation - which makes sense I guess, given that it is stateful on the server)

**Reply by rockfordlhotka** [reply]
@dazinator knows more about this than I do - perhaps he can help provide insight.

**Reply by dazinator** [reply]
> Is there some place we're supposed to add code to create the scope instead of letting Csla do it for use in a remote data portal host application (asp.net), and then dispose it?

Yes, if your remote application host is not asp.net core then you'll need to add code to your incoming request pipeline / processing to grab your platforms IServiceProvider and create a scope around processing of a request. For example for an OWIN based host you can add this logic to the OWIN middleware pipeline.

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka I did see that, but it wasn't clear to me if that would solve our issue or not.  When you say aspnet, do you mean .net framework or netcore?  We're still on .net framework.

**Reply by dazinator** [reply]
If you are using legacy asp.net as a host you'll have to remind me - does legacy aspnet even have an IServiceProvider? Or do you build your own?

**Reply by ajohnstone-ks** [reply]
@dazinator You can get IServiceProvider by adding some nuget packages, but I don't think asp.net FW will know anything about it.

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka So I think I have some things to try for the remote data portal hosted in asp.net framework, but I'm still not clear on where I would need to look to fix things in the Windows service.

Basically the service spins up a bunch of threads, and each of those threads will use a business object (typically a command) to do some work.  This is all done in local data portal mode, so client & server are the same physical tier.  I think this could be an issue on the desktop client app as well, but AFAIK we don't actually create an disposables in any logical server side code on the client.

It seems like something would need to happen soon after the transition to the logical server side code is made, but before Csla actually invokes the Create/Insert/Update/Delete/Fetch methods, and then dispose the scope before the transition back to the logical client side code. 

Any tips on where to start looking for this?

**Comment by dazinator** [comment]
@ajohnstone-ks 
See this: https://stackoverflow.com/questions/63017117/asp-webforms-with-system-web-http-api-controllers-constructor-dependency-inject

You'd need to be doing something similar to that to ensure an IServiceScope scope was created per request. You'd then need to call the csla application context extension method for setting the current scoped service provider to your ServiceScopes IServiceProvider at the start of the request - if this isn't set, csla will use the default IServiceProvider which I can see you are setting but this won't be created / disposed per request.

**Reply by dazinator** [reply]
@ajohnstone-ks 
Without creating a scope per request and telling csla what it is, csla will be falling back to the default service provider which will cause you issues where services won't be created / disposed per request, so this ties up with the issues you are reporting.

**Reply by ajohnstone-ks** [reply]
Yes, it looks like Csla will get a new scope created, but its not disposing it, so it leaves the connection around.

**Reply by rockfordlhotka** [reply]
Maybe what's needed, in a 5.4.x patch, is to have a method in `ApplicationContext` to dispose (and clear) the current scoped provider. Then you could implement `IInterceptDataPortal` and call that method on complete?

This would enable @ajohnstone-ks to meet his needs I think.

@dazinator would _not_ use this capability, because you are controlling your own scope(s).

However, @dazinator, the current implementation probably doesn't work for you _at all_, because `ApplicationContext` only understands one scope at a time, so there's no real way to have multiple scopes for different forms, etc. Isn't that a problem?

**Reply by dazinator** [reply]
> so there's no real way to have multiple scopes for different forms, etc. Isn't that a problem?

I used the pattern in an xamarin android application and android models UI navigation as `stack` of screens. The current screen showing on your phone is the one on top of the stack. Therefore there is only ever one active screen at a time. Think Windows forms modal dialogs where this would also hold true. This equates to one active scope at a time.

However AFAIK Csla doesn't impose this constraint - there is no reason right now why you can't have multiple active scopes. When you set the current scope IServiceProvider into application context I beleive (havent checked in a while) that the context manager should use something like AsyncLocal to propagated that field. Therefore different threads can have different scopes. If you have a single thread and you are switching between active windows I grant you, you will want a different way of either modelling or managing your scopes to suit your architecture.

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka Well, what I'd really need is a way to check if scope is null or not; if it is, I can create it and put it into a field in the interceptor.  Then on complete if its not null dispose it.  

Putting the interceptor aside, ideally I know that we just switched from logical client to logical server, and be able to execute code then (on the server, if the client & server are on physically different machines).  Then I can create a scope that's existence is tied to the running of the logical server side code and cleanup just before that server code returns.

@dazinator Our WinForms app is an MDI app, so we have multiple forms open at the same time.  We actually started implementing an MVVM architecture into it based on the MVVM architecture we used for our Xamarin.Forms app, so aside from single page / MDI they work in a similar manner (navigation & dialog services, etc).   Aside from application startup which bootstraps the container, and a Page factory which uses the container to build Forms/Pages & their VMs, nothing in our application knows anything about scopes or containers.

**Comment by dazinator** [comment, marked as answer]
Hi,
When your remote data portal runs locally, it means the dependencies CSLA
will want to inject are from some scope within your existing application.
Due to this, your application has more say in how it arranges its DI scopes
- I dont think CSLA should take control of creating seperate DI scopes
within your local application  - one per local DP operation for example.

On a remote server, serving requests from multiple clients, there is a
natural "per request" scope for each client and so this makes sense- and
you configure your services lifetimes with this model in mind.

When there is no remote server however, and the DP runs locally in your own
application - your application needs to supply scopes to csla. For example
perhaps in your Windows forms application, you want each new window to run
in its own scope - so that when the window closes, any dependencies
injected into will be disposed (db context etc). Perhaps you have some
background jobs - each of those jobs can have its own scope so anything
activated within the job will used dependencies scoped to that job and they
will all be cleaned up when the job finishes. You tell csla what scope its
operating within, and that impacts which dependencies will be consumed by
Csla. This allows your application to control where it creates scopes so
you can create scopes that last longer than a single DP invocation..

Therefore with the windows forms example, you'd need to figure out:

 Do you want to use scopes? If not, csla can use your top level
IServiveProvider - but make sure you don't register any IDisposable
services because they won't ever be disposed.

For your background threads example, if each thread needs its own instances
if dependencies then consider creating a scope per thread and having that
thread do any DI from that scope.

Hope this helps

**Reply by rockfordlhotka** [reply]
Another possible solution, is that the server-side data portal already had a way to support DI before Microsoft built it into modern .NET. There are hooks in the server-side data portal _specifically so you can create and dispose a scope_, and those remain valid.

The [CustomActivator sample](https://github.com/MarimerLLC/csla/tree/master/Samples/CustomActivator) demonstrates how to use the two interfaces involved.

You probably don't actually need to implement a custom activator anymore, because you can now rely on the `Inject` attribute in the data portal.

But if you implement and use a `IInterceptDataPortal` you can initialize and dispose a scope in the `Initialize` and `Complete` methods. In the `Initialize` method, just make sure to put your new scope into `ApplicationContext` so it is used as the current scope in the call.

I suppose alternately, you could allow CSLA to create the scope itself, and then just dispose that scope in the `Complete` method? That might be worth a try as well.

**Reply by rockfordlhotka** [reply]
Probably though, what you should consider is this:

Somewhere in your code you've created an endpoint - some tcp or http listener - and that listener invokes `Csla.Server.DataPortal`.

For example, the [HttpPortalController does this](https://github.com/MarimerLLC/csla/blob/05af980ab13158211b65dfa21766adc05e98edbd/Source/Csla.Web.Mvc.Shared/Server/Hosts/HttpPortalController.cs#L261). This is the standard endpoint for the HttpProxy, so the most used scenario right?

The thing is, with HttpPortalController, at the very top level, the [PostAsync method](https://github.com/MarimerLLC/csla/blob/05af980ab13158211b65dfa21766adc05e98edbd/Source/Csla.Web.Mvc.Shared/Server/Hosts/HttpPortalController.cs#L48) is what's invoked. Notice that it is virtual.

That means that in _your controller code_ that you put in your `Controllers` folder of your web site, you can override `PostAsync` - and that means you can create a scope, call the base `PostAsync` method, then dispose the scope.

In your service implementation you have some rough equivalent - each inbound network request is handled by some code that you wrote. Or you _could handle_.

**Reply by ajohnstone-ks** [reply]
@dazinator I don't think the application, once the container is bootstrapped, should really know anything at all about service scopes or the DI container.  There are some valid cases for a service locator pattern I think, but I prefer to save those for a last resort situation.  

Moving from a service to winforms, a button event handler would need code to create & dispose a scope, depending on whether or not the data portal method is a RunLocal or not.  A "UI" developer (winforms or windows service) shouldn't need to look at a BO to see if it can run locally or not and then implement scope management or not based on that knowledge.

@rockfordlhotka I'm looking at the IInterceptDataPortal, I think that will solve my issue.  Our service doesn't listen for any network connections; the threads spun up either fire on intervals via the Threading.Timer, or they a waiting on a Queue<T>, and wake up when the producer adds messsages and Pulses them to wake them up.  

I think its probably best of the business layer manages its own scopes (behind the scences); the problem with letting aspnet manage the scope is that you leak the database abstractions into the logical client side, and you get weird issues because the connection isn't closed until the end of the request (unknown to the controller developer).  It also can hold the connection open longer than it needs to be.

I think the only concern I have with the IInterceptDataPortal is that if in the server side business code it calls another root object (like a command) to do some other work, the command object will get its own intercepter object and start a new scope and thus get a new db connection.  I don't really see a good way around that though, because the getter for ApplicationContext.ServiceProviderScope is internal, so I can't check if one already exists.

**Reply by rockfordlhotka** [reply]
fwiw, the getter is internal on purpose - really we shouldn't be maintaining a reference to a service provider or scoped provider _at all_, but it is necessary to implement method parameter injection. I chose to make the getter non-public to avoid me or anyone else "cheating' by using the provider - thus falling into a bad habit.

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka Understood and I think that's valid; its just tripping me up in my particular scenario. 

I do think there should be some way to have csla create a scope after transitioning from logical client to logical server, and disposing it before the transition back.  It seems like most of the things people would want injected into the server side data portal code would need to be disposed, and would expect it to happen before client code is running again.

**Reply by dazinator** [reply]
@ajohnstone-ks 
Im not sure that its a safe assumption for Csla to create a new scope per csla operation. If you register scoped services it would mean a new instance being created and disposed per csla DP operation, and no way to make a scope last multiple DP operations. If you want to use scoped disposable dependencies, you need to have some higher level infrastructure create and dispose of the scope, and csla can operate within it.

**Reply by rockfordlhotka** [reply]
I tend to agree with @ajohnstone-ks though, because the whole point of the server-side data portal is that each request is an independent unit of work.

If the data portal is hosted in aspnetcore for example, each request _will_ have its own scope, because aspnetcore will see each request independently and will provide that scope. There is no provision for n concurrent requests that are dependent on each other.

@ajohnstone-ks in terms of there being such support, it does exist - that is exactly what the `IInterceptDataPortal` mechanism is designed to provide, in the case that you don't otherwise have code (like a controller) where you can more easily provide the implementation.

To maintain _your_ reference to the scoped provider instance, you can use `Csla.ApplicationContext.LocalContext` to store that reference, thus allowing you to 

1. Create the instance
2. Set the instance for use by CSLA
3. Store the instance in `LocalContext`

Then later

1. Retrieve the instance from `LocalContext`
2. Dispose the instance

**Reply by dazinator** [reply]
Just some notes, I hope they come accross as helpful but possibly not what you want to hear!

When you chose to register a dependency as "scoped" you must have a clear idea of that scope refers to in your application. When I register a service as scoped its because I expect to create / have scopes somewhere in my application. For an aspnet core application this is obvious - its the request or sometimes a background job, or sometimes I create a scope on startup so I can use EFCore DbContext to apply migrations etc. For windows forms or wpf apps this is less obvious, registering a dependency as scoped will exhibit the same behavior as registering it as singleton - if you aren't creating / using any scopes and just resolving it from root scope all the time. If you do create a scope per window, or a scope for different regions of the application, then registering services as scoped starts to mean something. If you aren't creating scopes in your application then registering something as scoped means nothing. If I create a scope for a background job, I dont want CSLA to create several of its own within that as it renders my applications concept of a DI scope useless. I want csla to honour my scope whether thats the current request or something else or even just root scope if I set things up to work from a single root scope.
The added complication is that you shouldn't register transient IDisposables in the root scope - because they will keep building up in the internal tracking list of the root scope and never be disposed. My take on a winforms or wpf application that wanted to use DI locally would be something like this:

1. Register dbcontext and others as scoped or transient.
2. Register forms / pages as transients to allow DI
3. When navigating to a new form, create a new DI scope, resolve the window / form instance from that scope, set the csla current scoped provider to that scoped service provider, and dispose of any old scope for the previous window that just closed.

This infrastructure code could be put in a central place. This would have the effect that:

1. You can use DI with forms / pages.
2. Scoped dependencies only live as long as the form. This means they get cleaned up as you navigate around the app, but also a scope now means something in your application - any thing you register as scoped typically means "scoped to the lifetime of a window / form"

Thats one fairly artifical but workable way and is a way I have used before with xamarin apps and csla.

If you don't like that definition of a scope, you can either try to make dependencies work safely from a singular root scope - although sadly injecting DbContexts that way (registered as scoped or transient) will lead to problems - so if you go that root you'd have to create some other service or factory to ensure you db contexts were created and disposed accordingly - Csla has DbContext manager for example.

**Reply by rockfordlhotka** [reply]
That makes sense @dazinator - thank you for the great explanation.

I think what is important here, is that we all understand that the server-side data portal (local or remote) IS ABSOLUTELY AN ATOMIC OPERATION. The cslanet framework goes to serious effort to ensure that, even when the server-side data portal is running locally, that it is isolated from the caller and acts as an atomic, remote operation.

As a result, _in the context of the server-side data portal_ I think it must be valid to create an use a scope. That is what will happen if the server-side data portal is running in aspnetcore, and thus that is what must happen if it runs in local mode. There must be parity of behavior - which translates to location transparency, and ensures that someone can simply switch their config from local to remote and things keep working.


Given that background, I fully appreciate what you are saying about client-side code in WinForms, WPF, Blazor, etc. You should control your own scopes within that client-side code!!

The server-side data portal though, that is a whole other thing, and (even when configured for local mode) is _logically_ not part of your client-side environment.

**Reply by dazinator** [reply]
@rockfordlhotka 
> though, because the whole point of the server-side data portal is that each request is an independent unit of work.

When the server side data portal runs on a remote host (not in the same local process as the client) - this issue is solved - you do get natural unit of work pattern by having a DI scope around each request on the remote host. However you get something else too..  the services registered for DI on the remove host are purely for CSLA data portal operations.  They aren't the same services that are registered for DI in the client application.. the remote host isn't "sharing" the same DI container as the client and doesn't need to know therefore what DI scopes are in play on the client - its irrelevant. For example you don't register an EFCore DBcontext directly in the startup of the client application as a scoped service, and expect Csla to use that same instance on the remote server. In fact you only need to register that dependency on the remote host and not in the local application as far as Csla is concerned.

Now switch to the mode where there is no remote host. Notice the difference now: csla doesn't have its own "remote dependencies" to satisfy DP operations with. Its actually sharing the client applications dependencies and reusing those instances where appropriate as dictated by the DI scopes of the application. I maintain that the application needs to be in control of its own scopes in this scenario. If csla wants to consume application dependencies to enable DI it should be aware / honour the scopes of the application in which those dependencies exist and not create new ones.. 

If however we were able to configure a seperate ServiceCollection and service provider for csla DP to use when running on the client, and this was completely seperate from your applications ServiceCollection / ServiceProvider- then I'd completely agree with everyone - the data portal can create scopes as it sees fit, and when registering services for "csla DI" registering a service as "scoped" now explicitly takes on the meaning of the csla DI "scope" which could mean per DP operation.

Thats my view, but I am open to be convinced otherwise.

**Reply by rockfordlhotka** [reply]
I think we're getting to the root of our misunderstanding 😉

This is _a core aspect of the server-side data portal_:

> There must be parity of behavior - which translates to location transparency, and ensures that someone can simply switch their config from local to remote and things keep working.

**Reply by rockfordlhotka** [reply]
And maybe what you say here

> If however we were able to configure a seperate ServiceCollection and service provider for csla DP to use when running on the client, and this was completely seperate from your applications ServiceCollection / ServiceProvider- then I'd completely agree with everyone

is the answer that will address the issue?

**Reply by rockfordlhotka** [reply]
This is, btw, a _great_ conversation to inform the CSLA 6 work item around DI: https://github.com/MarimerLLC/csla/issues/1738

**Reply by dazinator** [reply]
> is the answer that will address the issue?

I believe so. From a parity perspective - the slight difference is that when configuring the remote application, you can get away with configuring the main application container for use by the remote DP, because its assumed the sole purpose of the remote application is to host the DP. So at the application startup level you'd register CSLA dependencies as normal:

```
public void ConfigureServices(IServiceCollection services)
{
  services.AddDbContext<FooContext>();
}

```

Switching to the "non remote host" scenario - You could port this service registration code accross and into the client pretty much "as is" but you'd need to configure CSLA's own service collection / service provider for it's locally running DP to use, so you could imagine something like this:-

```
public void ConfigureServices(IServiceCollection services)
{
 // configuring my application dependencies here.
  services.AddTransient<DontCareAboutThis>();

  services.AddCsla().ConfigureServices(s=>{
       // we moved the csla registration code here!
       s.AddDbContext<FooContext>();
   }
}
```

Some quick things to point out:

1. There are two containers, so if you register a singleton in both your application services, and csla's services, by type (and not by instance) then you'll get two instances in memory.

This seems quite nice still, but if we did go this route, it would still be good to keep support for the model we have today which is:

1. You can hand the DP a scope to use, so that it won't create its own.

Keeping this feature just means that if people do want to "break parity" they can do so in return for potential performance gains by performing many DP operations within the context of their own scope. For example a background job that creates many users - saving each user would ordinarily perform a DP operation that in the new world, would create a new scope each time, and therefore create a new DbContext and dispose of it each time. This keeps parity with how the server works but doesn't allow the application to take advantage of the fact that `it is technically able to share memory with the DP running locally` - and so the application can potentially create its own scope, in order to reuse a single DbContext instance accross all of those operations.

**Reply by rockfordlhotka** [reply]
I like that @dazinator.

When I think "parity" I generally don't include config, especially of the server. There's no doubt that when you stand up an app server that you need to configure the server, and configure CSLA to run on the server.

The big issue with parity is that your non-config code shouldn't need changing, and the data portal should provide consistent behavior regardless of whether it is local or remote.

**Reply by ajohnstone-ks** [reply]
> If however we were able to configure a seperate ServiceCollection and service provider for csla DP to use when running on the client, and this was completely seperate from your applications ServiceCollection / ServiceProvider- then I'd completely agree with everyone - the data portal can create scopes as it sees fit, and when registering services for "csla DI" registering a service as "scoped" now explicitly takes on the meaning of the csla DI "scope" which could mean per DP operation.

This is exactly what the scopes are meant to provide; a new container which can be cleaned up when its use is over.  

Look at the code in the https://www.nuget.org/packages/Unity.Microsoft.DependencyInjection  When you call CreateScope, it creates a new ServiceProvider instance, and that instance gets a container created by calling container.CreateChild.

The child container doesn't inherit any objects registered as transients from its parent; instead it will create them new and store them.  When you dispose the child container, any disposable transients it created also get disposed.  

StructureMap has the same concept, I think it calls them nested containers though.  But the purpose is the same.

This is why I'm saying Csla should create a new scope on the transition to logical server side code, and then dispose it before moving back.  This will ensure that a call to `order = order.Save()` only keeps the connection open while the server side code is running, and that will happen consistently in local or remote data portal configurations.

@rockfordlhotka I think we are almost on the same page, but I think tying the scope Csla uses to the same scope an HttpController uses won't end well.  That was tried at a previous employer and we had unexpected behavior due to the sharing, and it kept the connection open longer than needed, which slowed the processing of requests.  This may not be evident until there's a high load though which makes troubleshooting more difficult.

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka Oh, duh.  I see what you're saying about LocalContext now.  My interceptor could look there for a scope, and then difference instances of the interceptor type would be able to all share one.  Ok... I think that will work for me.

I'm still thinking Csla should do scope management on the transition from server to client and back though; that will ensure parity between local and remote configurations.   It will also prevent others a nasty surprise as they move from using things like ConnectionManger to injection, where suddently the connection is being shared in situations where it wasn't previously. (Two root BOs used in a single action method, for example).

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka I implemented an interceptor to create & dispose a scope (but only if that instance of the interceptor created it to begin with).  Last night we deployed the fix in one of our windows services, and to one of the web servers hosting the data portal application host, and so far things are looking good.  We'll have to give it more time as the issue only popped up after about 24 hours, and we're not yet in our high load part of the day yet, but I have my fingers crossed.

There was an additional challenge in that I could get the service provider so I could use it to get a service scope factory, so the interceptor needed a static method to take the scope factor to use before it can do anything.  I just do this on application startup before setting the ApplicationContext.DefaultServiceProvider.

Thanks for pointing me in the right direction, much appreciated!

**Reply by dazinator** [reply]
@ajohnstone-ks 
> This is exactly what the scopes are meant to provide; a new container which can be cleaned up when its use is over.

> The child container doesn't inherit any objects registered as transients from its parent; instead it will create them new and store them. When you dispose the child container, any disposable transients it created also get disposed.

> StructureMap has the same concept, I think it calls them nested containers though. But the purpose is the same.

Yeah, there are multiple different DI container libraries, with different capabilities and features, however the so called "conforming" containers are the ones we care about - i.e they implement IServiceProvider and the scope factory etc.

Child containers are different from scopes / nested containers. When you create a new scope, the registrations are meant to be immutable - the new scope is purely a handle on a lifetime of services activated within that scope. Child containers can go a bit beyond that though and child containers are not provided by Microsoft.Extensions.DependencyInjection as it only provides scopes. Child Containers can inherit from a parent container but then allow you to modify registrations at the child container level. For example in a multi-tenant system you can register your common services in the parent container, and then have a child container per tenant with slightly different or customised registrations. I dont think we should be talking about child containers here though because this is all about scopes!

The motivation for my suggestion that the csla dataportal have its own IServiceCollection (and IServiceProvider) seperate from the one configured for the main application (when the data portal is also running locally) was mainly because - that is technically how the server side data portal works when it's running remotely - if you use a remote data portal- your client application has one set of services, and the remote data portal has another set of services - that's two distinct locations to register services and two distinct IServiceProviders. When there is no remote data portal in play and the data portal is instead now also operating within the local application I thought it could potentially be beneficial to keep these two sets of registrations (containers / IServiceProviders) distinct still - although that was just a thought. There are tradeoffs as always.

BTW: I'm glad you found a solution, there definitely can be some improvements made here for the future!

**Reply by ajohnstone-ks** [reply]
@dazinator Right... Unity calls that a child container.  It inherits all the registrations from its parent container, and the child container's purpose is to handle some of the registrations lifetime differently (except for things like Singleton, which should actually just be a singleton).  Transients created by a Unity child container will live and die with the container that created them.  You can use child containers for multi-tenant apps if you want, but one primary purpose is to allow you to dispose disposable transients.  If you look at the StructureMap documentation, the nested container concept was created explicitly for that purpose.

I'm talking about child containers because when you call IServiceProviderScopeFactory.CreateScope, the Microsoft implement which glues together Unity and their ServiceProvider framework will create a new child container.  When you dispose the scope, the container the ServiceProvider is referencing gets disposed (which would be the child container created).  I think that strongly suggests the connection between a ServiceProviderScope & Unit child container (or in StructureMap, nested container, but its been a while, so I might not be remembering the naming for SM correctly).  Other container frameworks will probably use other names, which makes this discussion a bit more confusing.  😄 

I understand where you coming from regarding a client vs server service collection, but I don't think that will actually provide much value.   I think you're going to have to fix registrations when you move from two tier to three tier no matter what.  And people will likely end up setting both client & server service collections to the same instance to avoid duplicating much of the same registrations; in our application, the server has almost all the same registrations as the client does, plus a few extra (for data access mainly).  There's only a few that are actually different.

The parity I'm most concerted about is the scope part of things.  Consider this MVC action method

```
[PreProcessingWhichUsesBOActionFilter]
public Task<ActionResult> CreateAccountAndPlaceOrder(dynamic accountAndOrderData) {
    var account = Account.NewAccount(accountAndOrderData.AccountData).Save();

   // Maybe other stuff happens here

    var order = Order.NewOrder(account, accountAndOrderData.OrderData).Save();

   return View(order);
}
```

If you were using ConnectionManager, a database connection would be opened & closed three times (assume the Creates don't use a DB connection), because there are three root BOs in use here (one's in the ActionFilter).  If people instead Inject their connection instead, its now unexpectedly shared between all three objects.  It will also leave the connection open until the very end of the request, instead of being closed immediately after the order.Save returns, which makes the connection unavailable for a longer period of time.  If the application with the above code is switch to a remote data portal, the connections are separate.  I think if you have the above code, then there should always be three connections open & closed whether you're running in 2-tier or 3-tier configurations.

**Reply by dazinator** [reply]
@ajohnstone-ks I am trying to follow, little confused over the following:

> if you were using ConnectionManager, a database connection would be opened & closed three times

And..

> If people instead Inject their connection instead

Using a connection manager to obtain a connection, and having a connection injected are two different actions so the behaviour will be different from one style to the other.

In the case of using a connection manager, its essentially in control of the lifetime of the connection, so if you use this on the server or in the client, the behaviour should be governed by the connection manager and so should be consistent in terms of "parity". You seem to be comparing using a connection manager to injecting a connection, and they are two different cases. When an IDisposable dependency is injected into a service, that service is not meant to take responsibility for the disposable of that service - this is instead governed by the container or 'scope' that the dependency lives in - when the scope is disposed the dependency will be disposed. Therefore when you are using DI to inject IDisposables into a class that class doesn't know how long that IDisposable will live for and whether it will be reused, and its dangerous to make assumptions about that. To your example, if the DI scope is at the request level and the connection the same connection is only disposed of at the end of the request, or if you chose to create seperate scopes per individual CSLA operation and therefore each csla class had its own IDisposable connection injected and then disposed - those are both valid use cases in my view and that is precisely the power and use case for creating DI scopes.

So for me this is mainly all about, at what point scopes should be created and should csla create them for you. In my view if CSLA creates a scope and disposes of a scope for every individual DP operation, this would seem to be more in line with how things work for a remote data  portal so I think I am on board with that. However I still think its really beneficial to be able do what is in the framework today, and that is to create your own scoped IServiceProvider and have the DP use that without creating individual "sub scopes" per DP operation- for example we actually wanted exactly what you mentioned - during a request we wanted the same instance of the DBContext to be used by the middleware, by the MVC Action / Controller, and then also by all the Csla DP operations throughout that request, and disposing of that instance at the end of the request. This meant everything had a consistent view of entities being tracked etc and was more efficient than creating several dbcontexts per request. We can do this today because we can allow the DP running locally, to use the aso.net core request scoped IServiceProvider and it doesnt create individual subscopes.

So I think it sounds like CSLA needs to allow for both scenarios here perhaps :-)

**Reply by dazinator** [reply]
@ajohnstone-ks 
I am thinking one solution might be..

1. Allow a user to set the current scoped SP in csla application context (we have this already in place today)
2. If csla detects that this is set, the DP will use it without creating a scope itself. (Which is what happens today)
3. If csla detects that it isn't set, csla can get the current IServiceProvider (it does this today), and create a scope itself (it doesn't do this bit), around the DP operation running locally, disposing of the scope one its complete. This would give you the behavior I think you are after, and the ability to control scopes more when needed.

I think its something like this that we need, open to other alternatives.
@rockfordlhotka is the best person to comment on how this kind of thing might be implemented. Would this work for you?

**Reply by rockfordlhotka** [reply]
That makes sense to me.

There could be a default interceptor that does this. And if you don't like it, you can replace the default interceptor.

This default interceptor would then dispose and clear the scoped provider when the data portal operation is complete right?

Also, this default interceptor would examine the current scoped provider value. If it is null then it creates a new one and disposes it on complete. If it is not null, then the interceptor would do nothing at all.

**Reply by dazinator** [reply]
> Also, this default interceptor would examine the current scoped provider value. If it is null then it creates a new one and disposes it on complete

Yep that sounds right to me. Creating a new one would involve:

1. Getting the current IServiceProvider. We have logic today for that which detects an active request and returns the one scoped to current request, if there is no request falls back to returning the root / application level IServiceProvider.
2. Once you have the IServiceProvider you can call the IServiceProvider.CreateScope() extension method to create a new IServiceScope. IServiceScope is IDisposable and should be kept private to the interceptor and disposed after the DP operation is complete, which disposes of its IServiceProvider.
3. The IServiceScope has a `ServiceProvider` property that is the IServiceProvider for this new scope - that can be set in Csla Application Context for the current scoped IServiceProvider - i.e the same location that the interceptor found was NULL initially. After the DP operation is complete the interceptor should set this back to NULL again.
4. The DP injection logic can continue to use the IServiceProvider set in application context. If none is set (which may be the case on a remote host) it would fall back to using the IServiceProvider from current request or if no request fallback to application level service provider.

So to recap, a service provider can be set in csla application context:

1. By user code
2. The interceptor if it was NULL @rockfordlhotka does this interceptor logic only run on the client right? I.e - when the server side DP is also running directly on the client?

On a remote host, we wouldn't want the interceptor to run and create a new scope, because for example - asp.net core will already have created a perfectly good scope around the incoming DP request anyway. For other hosts (non asp.net core) there may need to be logic added that sets the request scoped IServiceProvider into csla application context for the inbound request - this code would essentially be the same as the interceptor but on the remote server?

The DP logic that actually does the injection will use the IServiceProvider from the following chain of "fallbacks" from highest to lowest:
- The one set in Csla application context 
- The one found for the current request (this is done by detecting an active httpcontext.requestservices) which will work on aspnet core hosts.
- The one captured on application startup- i.e the main application level IServiceProvider.

**Reply by ajohnstone-ks** [reply]
> @ajohnstone-ks
> I am thinking one solution might be..
> 
> 1. Allow a user to set the current scoped SP in csla application context (we have this already in place today)
> 2. If csla detects that this is set, the DP will use it without creating a scope itself. (Which is what happens today)
> 3. If csla detects that it isn't set, csla can get the current IServiceProvider (it does this today), and create a scope itself (it doesn't do this bit), around the DP operation running locally, disposing of the scope one its complete. This would give you the behavior I think you are after, and the ability to control scopes more when needed.
> 
> I think its something like this that we need, open to other alternatives.
> @rockfordlhotka is the best person to comment on how this kind of thing might be implemented. Would this work for you?

If I am following, I think what you are saying is:

* current scoped SP = ApplicatoinContext.ServiceProviderScope (correct?)
* If ServiceProviderScope != null, do nothing
* If ServiceProviderScope == null, ServiceProviderScope = DefaultServiceProvider.GetRequiredService<IServiceScopeFactory>().CreateScope().ServiceProvider (for Csla 5.4.0)

And if #3 is happened, dispose the scope.

I would agree; I think the scope checking & creation should be done after moving from logical client to logical server, and cleanup just before returning to logical client.

That's actually exactly what I had been expecting Csla to do when we implemented the injection.  This would allow disposable resources needed in the logical server code to be disposed of before returning to the logical client side, which I think is a pretty reasonable expectation.  And if people can override that default behavior, they can do so.

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka @dazinator That's exactly what the interceptor I coded does; it was just a bit tricker b/c I couldn't look at the values for DefaultServiceProvider/ServiceProviderScope, so I had to have a static method to configure it on startup and track the created scope separately.

Very much appreciate both of you discussing this!

**Reply by ajohnstone-ks** [reply]
> On a remote host, we wouldn't want the interceptor to run and create a new scope, because for example - asp.net core will already have created a perfectly good scope around the incoming DP request anyway.

I think a scope should still be created & disposed even in a remote host.  

My reasoning is that there is time between when Csla would call Complete on an interceptor and the end of the asp.net request and then dispose the scope.   

For most applications this will not be an issue, but under heavy load the connection being kept open & unavailable to other requests will matter, even if only for fractions of a second.  

@rockfordlhotka I've created an issue for this as requested https://github.com/MarimerLLC/csla/issues/1996

**Reply by rockfordlhotka** [reply]
Thank you for creating the issue @ajohnstone-ks. The great thing about this solution with the interceptor, is that you _can_ use the interceptor on the server if you'd like - it'll be a configuration choice between a no-op interceptor and the per-call scope interceptor, regardless of where the server-side data portal is running.

btw, can you share the code from your custom interceptor? Perhaps in the issue instead of here, as it would be useful for reference in the future.

**Reply by ajohnstone-ks** [reply]
@rockfordlhotka Yes, I fully agree this is a great solution as it is fully customizable.  

I've added code similar to the one I implemented to the issue; its a bit easier because you can get DSP & SPS from within Csla and I've added some points to allow someone to override Initialize & Complete if they want to add behavior in addition to keeping the scope behavior.  I've also updated the ticket to reflect that it shouldn't be configured by default for a remote application server, since that doesn't seem to be desired there.
